config/src/network_id.rs

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct NetworkInfo {
+    name: String,
+}
+
+/// A representation of the network being used in communication.
+/// There should only be one of each NetworkId used for a single node, and handshakes should verify
+/// that the NetworkId being used is the same during a handshake, to effectively ensure communication
+/// is restricted to a network.  Network should be checked that it is not the `DEFAULT_NETWORK`
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+#[serde(tag = "type")]
```

Comment:
This option is likely the cause of the breakage in the `generate-format` unit test. It would not work with LCS in first place.

Any reason for using this? Is this meant to be serialized both in JSON and LCS? (then I don't have a solution but to accept a slightly uglier JSON)

Reply:
Can we eliminate the notion of "tag" from LCS, I don't think we need to serialize / deserialize this field... I don't know why serde would do this to us :P

Reply:
OK so if we really want to share the definition of `NetworkId` between TOML and LCS, here is a solution:
```
#[derive(Clone, Debug, Eq, Hash, PartialEq)]
pub enum NetworkId {
    Validator,
    Public,
    Private(NetworkInfo),
}

#[derive(Serialize, Deserialize)]
#[serde(tag = "type", rename = "NetworkId")]
enum HumanReadableNetworkId {
    Validator,
    Public,
    Private(NetworkInfo),
}

#[derive(Serialize, Deserialize)]
#[serde(rename = "NetworkId")]
enum NonHumanReadableNetworkId {
    Validator,
    Public,
    Private(NetworkInfo),
}

impl Serialize for NetworkId {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer
    {
        if serializer.is_human_readable() {
            match self {
                NetworkId::Validator => HumanReadableNetworkId::Validator.serialize(serializer),
                NetworkId::Public => HumanReadableNetworkId::Public.serialize(serializer),
                NetworkId::Private(info) => HumanReadableNetworkId::Private(*info).serialize(serializer),
            }
        } else {
            match self {
                NetworkId::Validator => NonHumanReadableNetworkId::Validator.serialize(serializer),
                NetworkId::Public => NonHumanReadableNetworkId::Public.serialize(serializer),
                NetworkId::Private(info) => NonHumanReadableNetworkId::Private(*info).serialize(serializer),
            }
        }
    }
}

impl<'de> Deserialize<'de> for NetworkId {

    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>
    {
        if deserializer.is_human_readable() {
            match HumanReadableNetworkId::deserialize(deserializer)? {
                HumanReadableNetworkId::Validator => Ok(NetworkId::Validator),
                HumanReadableNetworkId::Public => Ok(NetworkId::Public),
                HumanReadableNetworkId::Private(info) => Ok(NetworkId::Private(info)),
            }
        } else {
            match NonHumanReadableNetworkId::deserialize(deserializer)? {
                NonHumanReadableNetworkId::Validator => Ok(NetworkId::Validator),
                NonHumanReadableNetworkId::Public => Ok(NetworkId::Public),
                NonHumanReadableNetworkId::Private(info) => Ok(NetworkId::Private(info)),
            }
        }
}
```

Obviously, it's not ideal but it works. If this happens often in the future, we could think of some procedural macros to help.